KIP-374: Add the missing '--help' option to Kafka commands (KAFKA-7418)

mrsrinivas opened the pull request:
Changes made as part of this [KIP-374](https://cwiki.apache.org/confluence/x/FgSQBQ) and [KAFKA-7418](https://issues.apache.org/jira/browse/KAFKA-7418)
 - Checking for empty args or help option in command file to print Usage
 - Added new class to enforce help option to all commands
 - Refactored few lines (ex `PreferredReplicaLeaderElectionCommand`) to
   make use of `CommandDefaultOptions` attributes.
 - Made the changes in help text wordings

Run the unit tests in local(Windows) few Linux friendly tests are failing but 
not any functionality, verified `--help` and no option response by running 
Scala classes, since those all are having `main` method.
 
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mrsrinivas committed:
KAFKA-7418: Add '--help' option to missing Kafka CLI commands (KIP-374)

mrsrinivas committed:
KAFKA-7418: Add '--help' option to missing Kafka CLI commands (KIP-374)

mrsrinivas committed:
Merge remote-tracking branch 'origin/KIP-374' into KIP-374

# Conflicts:
#	connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java
#	connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectStandalone.java

mrsrinivas commented:
[![Build Status](https://builds.apache.org/buildStatus/icon?job=kafka-pr-jdk8-scala2.11&build=17737)](https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/17737/) - JDK 8 and Scala 2.11  

[![Build Status](https://builds.apache.org/buildStatus/icon?job=kafka-pr-jdk11-scala2.12&build=677)](https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/677/) - JDK 11 and Scala 2.12 

retest this please, since test failures are not related to code changes in this PR.

cmccabe commented on core/src/main/scala/kafka/utils/CommandLineUtils.scala:
```diff
@@ -16,21 +16,44 @@
+    * @param commandOpts Acceptable options for a command
+    * @param message     Message to display on successful check
+    */
+  def checkHelpArgAndPrintUsageAndDie(commandOpts: CommandDefaultOptions, message: String) = {
```
The name is kind of confusing.  How about "printHelpAndExitIfNeeded" or similar.

mrsrinivas replied:
I changed the method name to `printHelpAndExitIfNeeded`

cmccabe commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -157,16 +157,10 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
-        parser.printHelpOn(System.out)
-        sys.exit(0)
-      }
+      if (args.length == 0 || options.has(helpOpt))
```
We should use a utility function here rather than repeating the "check args or check for help flag" logic.

mrsrinivas replied:
created options subclass for following command classes so that `CommandLineUtils#printHelpAndExitIfNeeded` method can be used for help check.

- ZkSecurityMigrator
- DumpLogSegments
- MirrorMaker

cmccabe commented:
Thanks, @mrsrinivas.  I left a comment about the helper function names.  Looks good aside from that.

mrsrinivas committed:
KAFKA-7418: Addressed review comments
 - changed method name as suggested
 - Unified config options by adding subclass so that help
 check common now

mrsrinivas committed:
KAFKA-7418: Addressed review comments
 - changed method name as suggested
 - Unified config options by adding subclass so that help
 check common now

mrsrinivas commented:
ping @cmccabe  @omkreddy @guozhangwang @ijuma

mrsrinivas commented:
[![Build Status](https://builds.apache.org/buildStatus/icon?job=kafka-pr-jdk8-scala2.11&build=17801)](https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/17801/) - JDK 8 and Scala 2.11  

[![Build Status](https://builds.apache.org/buildStatus/icon?job=kafka-pr-jdk11-scala2.12&build=739)](https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/739/) - JDK 11 and Scala 2.12 

retest this please, since test failures are not related to code changes in this PR.

omkreddy commented:
Can we also update LogDirsCommand.scala and StreamsResetter.java files?

guozhangwang commented on core/src/main/scala/kafka/admin/PreferredReplicaLeaderElectionCommand.scala:
```diff
@@ -109,13 +92,34 @@ object PreferredReplicaLeaderElectionCommand extends Logging {
+
+    options = parser.parse(args: _*)
+
+    def checkArgs() {
```
nit: this function seems an overkill to me, we can just call `CommandLineUtils.checkRequiredArgs(parser, options, zkConnectOpt)` as we did in other classes.

guozhangwang commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -72,8 +72,7 @@ object ConfigCommand extends Config {
 
-      if (args.length == 0)
-        CommandLineUtils.printUsageAndDie(opts.parser, "Add/Remove entity config for a topic, client, user or broker")
+      CommandLineUtils.printHelpAndExitIfNeeded(opts, "Add/Remove entity config for a topic, client, user or broker")
```
Add / remove -> manipulate and describe

guozhangwang reviewed (commented):
This is a nice cleanup! I made a pass over it except the MM class.

mimaison commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -53,8 +53,7 @@ object AclCommand extends Logging {
 
-    if (opts.options.has(opts.helpOpt))
-      CommandLineUtils.printUsageAndDie(opts.parser, "Usage:")
+    CommandLineUtils.printHelpAndExitIfNeeded(opts, "This tools helps to manage acls on kafka.")
```
nit: This _tool_ helps

mimaison replied:
It still says "This tools helps". It should be "This tool helps"

mimaison commented on core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala:
```diff
@@ -72,10 +70,11 @@ object BrokerApiVersionsCommand {
     checkArgs()
 
     def checkArgs() {
+      CommandLineUtils.printHelpAndExitIfNeeded(this, "Command to retrieving broker version information.")
```
Should we use a similar description format for all these tools? AclCommand.scala uses `This tools helps ...` while this one uses `Command to retrieving ...`. I personally prefer the former but can you pick one and apply it everywhere?

mimaison reviewed (changes requested):
Thanks for the updates, overall logic looks good.
I've left a couple of comments regarding small cleanups

mrsrinivas committed:
Merge remote-tracking branch 'origin/KIP-374' into KIP-374

# Conflicts:
#	core/src/main/scala/kafka/admin/AclCommand.scala
#	core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala
#	core/src/main/scala/kafka/admin/ConfigCommand.scala
#	core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala
#	core/src/main/scala/kafka/admin/DelegationTokenCommand.scala
#	core/src/main/scala/kafka/admin/DeleteRecordsCommand.scala
#	core/src/main/scala/kafka/admin/PreferredReplicaLeaderElectionCommand.scala
#	core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala
#	core/src/main/scala/kafka/admin/TopicCommand.scala
#	core/src/main/scala/kafka/tools/ConsoleConsumer.scala
#	core/src/main/scala/kafka/tools/ConsoleProducer.scala
#	core/src/main/scala/kafka/tools/ConsumerPerformance.scala
#	core/src/main/scala/kafka/tools/DumpLogSegments.scala
#	core/src/main/scala/kafka/tools/MirrorMaker.scala

mrsrinivas commented:
@omkreddy @guozhangwang @mimaison, thanks for review feedback.

I addressed all comments in latest commit.

mrsrinivas committed:
Merge branch 'trunk' into KIP-374

mimaison reviewed (approved):
Thanks for the quick updates. LGTM

mrsrinivas commented:
ping @omkreddy @guozhangwang @cmccabe

guozhangwang commented:
LGTM. I'd leave it to @omkreddy for the final review and merge.

omkreddy reviewed (approved):
@mrsrinivas Thanks for the PR. LGTM. Thanks for your patience!.

ijuma commented on core/src/main/scala/kafka/tools/DumpLogSegments.scala:
```diff
@@ -495,4 +394,104 @@ object DumpLogSegments {
+      "transaction metadata from the __transaction_state topic.")
+    options = parser.parse(args : _*)
+
+    def checkArgs() = {
```
I am trying to resolve some conflicts in one of my PRs and I am a bit puzzled why all the logic of this command is in a method called `checkArgs`. Can someone clarify, please?

mrsrinivas replied:
Most of the places, the logic is using the local variables. Since the local variables moved into new `DumpLogSegmentsOptions` then ended up binding the logic into one `checkArgs` method. 

 Alternately, variables reference could be modified with `opts.` prefix by keeping the logic there. 

But, I ended up doing this to keep consistency with other commands like `AclCommand.scala` implementation and thought of it's relevant to place validation logic in new class since it's there.

ijuma replied:
I think this approach is confusing since `checkArgs` is actually running the command. I will fix this in my PR. Thanks for the explanation.
